src/ucs/datastruct/callbackq.c

```diff
@@ -7,322 +7,516 @@
+
+    /* Disable this proxy if no more work to do */
+    if (!priv->fast_remove_mask && (priv->num_slow_elems == 0)) {
+        ucs_callbackq_disable_proxy(cbq);
     }
 
     ucs_callbackq_leave(cbq);
+}
+
+ucs_status_t ucs_callbackq_init(ucs_callbackq_t *cbq)
+{
+    ucs_callbackq_priv_t *priv = ucs_callbackq_priv(cbq);
+    unsigned idx;
+
+    for (idx = 0; idx < UCS_CALLBACKQ_FAST_COUNT; ++idx) {
+        ucs_callbackq_elem_reset(cbq, &cbq->fast_elems[idx]);
+    }
+
+    ucs_spinlock_init(&priv->lock);
+    priv->slow_elems        = NULL;
+    priv->num_slow_elems    = 0;
+    priv->max_slow_elems    = 0;
+    priv->slow_idx          = 0;
+    priv->slow_proxy_id     = UCS_CALLBACKQ_ID_NULL;
+    priv->fast_remove_mask  = 0;
+    priv->num_fast_elems    = 0;
+    priv->free_idx_id       = UCS_CALLBACKQ_ID_NULL;
+    priv->num_idxs          = 0;
+    priv->idxs              = NULL;
     return UCS_OK;
 }
 
-static void ucs_callbackq_slow_path_cb(void *arg)
+void ucs_callbackq_cleanup(ucs_callbackq_t *cbq)
 {
-    ucs_callbackq_t *cbq = arg;
-    ucs_callbackq_slow_elem_t *elem, *tmp_elem;
+    ucs_callbackq_priv_t *priv = ucs_callbackq_priv(cbq);
 
-    ucs_callbackq_enter(cbq);
+    ucs_callbackq_disable_proxy(cbq);
 
-    /* TODO avoid locks on every iteration. For instance implement
-     * lazy remove with additional list of elements to be removed. */
-    ucs_list_for_each_safe(elem, tmp_elem, &cbq->slow_path, list) {
-        ucs_callbackq_leave(cbq);
-        elem->cb(elem);
-        ucs_callbackq_enter(cbq);
+    if ((priv->num_fast_elems) > 0 || (priv->num_slow_elems > 0)) {
+        ucs_warn("%d fast-path and %d slow-path callbacks remain in the queue",
+                 priv->num_fast_elems, priv->num_slow_elems);
     }
 
-    ucs_callbackq_leave(cbq);
+    ucs_callbackq_array_free(priv->slow_elems, sizeof(*priv->slow_elems),
+                             priv->max_slow_elems);
+    ucs_callbackq_array_free(priv->idxs, sizeof(*priv->idxs), priv->num_idxs);
 }
 
-void ucs_callbackq_add_slow_path(ucs_callbackq_t *cbq,
-                                 ucs_callbackq_slow_elem_t* elem)
+int ucs_callbackq_add(ucs_callbackq_t *cbq, ucs_callback_t cb, void *arg,
+                      unsigned flags)
 {
-    ucs_status_t status;
+    ucs_callbackq_priv_t *priv = ucs_callbackq_priv(cbq);
+    int id;
 
     ucs_callbackq_enter(cbq);
 
-    /* Note: multiple calls with the same element are not allowed */
-    ucs_list_add_tail(&cbq->slow_path, &elem->list);
-    status = ucs_callbackq_add_safe(cbq, ucs_callbackq_slow_path_cb, cbq);
-    ucs_assert_always(status == UCS_OK);
+    ucs_trace_func("cbq=%p cb=%s arg=%p flags=%u", cbq,
+                   ucs_debug_get_symbol_name(cb), arg, flags);
+
+    if ((flags & UCS_CALLBACKQ_FLAG_FAST) &&
+        (priv->num_fast_elems < UCS_CALLBACKQ_FAST_MAX))
+    {
+        id = ucs_callbackq_add_fast(cbq, cb, arg, flags);
+    } else {
+        id = ucs_callbackq_add_slow(cbq, cb, arg, flags);
```

Comment:
So, callbacks are added to the fast path according to initialization sequence. For instance if we have > than 8 transports, only the first 7 will be added to FP (regardless of their priorities). Maybe we may consider adding some sort of priorities to make sure that certain cb will be added to FP in any case?